File: crates/book-forge-converter/src/url_tools.rs
use url::Url;

pub(crate) fn same_origin(left: &Url, right: &Url) -> bool {
    left.scheme() == right.scheme()
        && left.host_str() == right.host_str()
        && left.port_or_known_default() == right.port_or_known_default()
}

pub(crate) fn normalize_page_url(url: &Url) -> String {
    let mut normalized = url.clone();
    normalized.set_fragment(None);
    normalize_default_page_path(&mut normalized);
    normalized.to_string()
}

pub(crate) fn normalize_resource_url(url: &Url) -> String {
    let mut normalized = url.clone();
    normalized.set_fragment(None);
    normalized.to_string()
}

pub(crate) fn url_without_fragment(url: &Url) -> String {
    let mut normalized = url.clone();
    normalized.set_fragment(None);
    normalized.to_string()
}

pub(crate) fn default_prefix_for(start_url: &Url) -> String {
    let mut prefix = start_url.clone();
    prefix.set_query(None);
    prefix.set_fragment(None);

    let path = prefix.path();
    let directory = if path.ends_with('/') {
        path.to_string()
    } else {
        match path.rsplit_once('/') {
            Some(("", _)) => "/".to_string(),
            Some((directory, _)) => format!("{directory}/"),
            None => "/".to_string(),
        }
    };

    prefix.set_path(&directory);
    prefix.to_string()
}

fn normalize_default_page_path(url: &mut Url) {
    let path = url.path();
    let mut normalized = path.to_string();

    for suffix in ["/index.html", "/index.htm"] {
        if normalized.ends_with(suffix) {
            let keep = normalized.len() - suffix.len() + 1;
            normalized.truncate(keep);
            break;
        }
    }

    if normalized.is_empty() {
        normalized.push('/');
    }

    if !normalized.ends_with('/') {
        let last_segment = normalized.rsplit('/').next().unwrap_or_default();
        if !last_segment.contains('.') {
            normalized.push('/');
        }
    }

    url.set_path(&normalized);
}

File: converter/urltools.go
package converter

import (
	"net/url"
	"strings"
)

func parseURL(raw string) (*url.URL, error) {
	return url.Parse(raw)
}

func sameOrigin(left, right *url.URL) bool {
	leftPort := left.Port()
	rightPort := right.Port()
	if leftPort == "" {
		leftPort = defaultPortForScheme(left.Scheme)
	}
	if rightPort == "" {
		rightPort = defaultPortForScheme(right.Scheme)
	}
	return strings.EqualFold(left.Scheme, right.Scheme) &&
		strings.EqualFold(left.Hostname(), right.Hostname()) &&
		leftPort == rightPort
}

func defaultPortForScheme(scheme string) string {
	switch strings.ToLower(scheme) {
	case "http":
		return "80"
	case "https":
		return "443"
	default:
		return ""
	}
}

func normalizePageURL(u *url.URL) string {
	nu := cloneURL(u)
	nu.Fragment = ""
	nu.RawQuery = ""
	normalizeDefaultPagePath(nu)
	return nu.String()
}

func normalizeResourceURL(u *url.URL) string {
	nu := cloneURL(u)
	nu.Fragment = ""
	return nu.String()
}

func urlWithoutFragment(u *url.URL) string {
	nu := cloneURL(u)
	nu.Fragment = ""
	return nu.String()
}

func defaultPrefixFor(startURL *url.URL) string {
	prefix := cloneURL(startURL)
	prefix.RawQuery = ""
	prefix.Fragment = ""

	path := prefix.Path
	if !strings.HasSuffix(path, "/") {
		if idx := strings.LastIndexByte(path, '/'); idx > 0 {
			path = path[:idx+1]
		} else {
			path = "/"
		}
	}
	prefix.Path = path
	return prefix.String()
}

func cloneURL(u *url.URL) *url.URL {
	nu := *u
	nu.User = nil
	return &nu
}

func normalizeDefaultPagePath(u *url.URL) {
	path := u.Path
	normalized := path

	for _, suffix := range []string{"/index.html", "/index.htm"} {
		if strings.HasSuffix(normalized, suffix) {
			normalized = normalized[:len(normalized)-len(suffix)+1]
			break
		}
	}

	if normalized == "" {
		normalized = "/"
	}

	if !strings.HasSuffix(normalized, "/") {
		lastSlash := strings.LastIndexByte(normalized, '/')
		lastSegment := normalized[lastSlash+1:]
		if !strings.Contains(lastSegment, ".") {
			normalized += "/"
		}
	}

	u.Path = normalized
}
